File: C#/CorsTest.cs
namespace SecurityTester.Tests;

public class CorsTest
{
    private static readonly string[] Endpoints =
        { "", "/api", "/api/users", "/api/data" };

    public static async Task<TestResult> RunAsync(string target, HttpClient client, SiteMap? siteMap = null)
    {
        var result     = new TestResult { Name = "CORS Misconfiguration" };
        var baseUrl    = target.TrimEnd('/');
        var evilOrigin = "https://evil-attacker.com";

        foreach (var endpoint in Endpoints)
        {
            var url = $"{baseUrl}{endpoint}";
            try
            {
                var request = new HttpRequestMessage(HttpMethod.Options, url);
                request.Headers.Add("Origin", evilOrigin);
                request.Headers.Add("Access-Control-Request-Method", "GET");

                    var res = await client.SendAsync(request);
                    await Task.Delay(100);
                    if (!res.Headers.TryGetValues("access-control-allow-origin", out var acao)) continue;

                var origin = string.Join("", acao);
                res.Headers.TryGetValues("access-control-allow-credentials", out var acac);
                var credentials = string.Join("", acac ?? Array.Empty<string>());

                if ((origin == "*" || origin == evilOrigin) && credentials == "true")
                {
                    result.Passed = false;
                    result.Issues.Add(new SecurityIssue
                    {
                        Title    = "CRITICAL CORS — Any Origin + Credentials Allowed",
                        Location = url,
                        Detail   = $"Access-Control-Allow-Origin: {origin}\nAccess-Control-Allow-Credentials: true",
                        Severity = "CRITICAL",
                        Risk     = "Any website can make authenticated requests to your API — reading private data or taking over accounts.",
                        Fixes    = new()
                        {
                            new() { Label = "C# / ASP.NET Core — whitelist specific origins",
                                    Code  = "// In Program.cs:\nbuilder.Services.AddCors(options =>\n{\n    options.AddPolicy(\"SafeCors\", policy =>\n    {\n        policy.WithOrigins(\"https://yoursite.com\", \"https://app.yoursite.com\")\n              .AllowCredentials()\n              .AllowAnyMethod()\n              .AllowAnyHeader();\n    });\n});\napp.UseCors(\"SafeCors\");" },
                            new() { Label = "Never combine wildcard with credentials",
                                    Code  = "// ❌ EXTREMELY DANGEROUS\npolicy.AllowAnyOrigin().AllowCredentials();\n\n// ✅ SAFE\npolicy.WithOrigins(\"https://yoursite.com\").AllowCredentials();" }
                        }
                    });
                }
                else if (origin == "*")
                {
                    result.Issues.Add(new SecurityIssue
                    {
                        Title    = "CORS Allows All Origins (Wildcard)",
                        Location = url,
                        Detail   = "Access-Control-Allow-Origin: * — any website can read this API's responses.",
                        Severity = "MEDIUM",
                        Risk     = "Fine for public APIs. Dangerous if any sensitive data is returned.",
                        Fixes    = new()
                        {
                            new() { Label = "Restrict to known origins if response has sensitive data",
                                    Code  = "policy.WithOrigins(\"https://yoursite.com\");" }
                        }
                    });
                }
                else if (origin == evilOrigin)
                {
                    result.Passed = false;
                    result.Issues.Add(new SecurityIssue
                    {
                        Title    = "CORS Reflects Arbitrary Origin",
                        Location = url,
                        Detail   = $"Server reflected evil origin: Access-Control-Allow-Origin: {origin}",
                        Severity = "HIGH",
                        Risk     = "Server blindly allows any origin — effectively wildcard but works with credentials too.",
                        Fixes    = new()
                        {
                            new() { Label = "Validate origin strictly before reflecting",
                                    Code  = "var allowed = new HashSet<string> { \"https://yoursite.com\" };\nif (allowed.Contains(request.Headers[\"Origin\"].ToString()))\n    response.Headers.Add(\"Access-Control-Allow-Origin\", request.Headers[\"Origin\"]);" }
                        }
                    });
                }
            }
            catch { }
        }

        return result;
    }
}

File: security-tester-vscode/tests/corsTest.js
const https = require("https");
const http  = require("http");
const { URL } = require("url");

const delay = ms => new Promise(resolve => setTimeout(resolve, ms));

const ENDPOINTS = ["", "/api", "/api/users", "/api/data"];
const EVIL_ORIGIN = "https://evil-attacker.com";

function fetchOptions(url, origin) {
  return new Promise((resolve) => {
    try {
      const parsed = new URL(url);
      const lib    = parsed.protocol === "https:" ? https : http;
      const opts = {
        method: "OPTIONS",
        hostname: parsed.hostname,
        port: parsed.port || (parsed.protocol === "https:" ? 443 : 80),
        path: parsed.pathname + parsed.search,
        headers: {
          "Origin": origin,
          "Access-Control-Request-Method": "GET",
        },
        rejectUnauthorized: false,
        timeout: 10000,
      };
      const req = lib.request(opts, (res) => {
        const acao = res.headers["access-control-allow-origin"] || "";
        const acac = res.headers["access-control-allow-credentials"] || "";
        res.resume();
        res.on("end", () => resolve({ acao, acac }));
      });
      req.on("error",   () => resolve(null));
      req.on("timeout", () => { req.destroy(); resolve(null); });
      req.end();
    } catch { resolve(null); }
  });
}

async function testCORS(target, siteMap) {
  const result  = { name: "CORS Misconfiguration", passed: true, issues: [] };
  const baseUrl = target.replace(/\/$/, "");

  for (const endpoint of ENDPOINTS) {
    const url = `${baseUrl}${endpoint}`;
    try {
      const res = await fetchOptions(url, EVIL_ORIGIN);
      if (!res) continue;
      await delay(100);

      const origin      = res.acao;
      const credentials = res.acac;

      if ((origin === "*" || origin === EVIL_ORIGIN) && credentials === "true") {
        result.passed = false;
        result.issues.push({
          title:    "CRITICAL CORS — Any Origin + Credentials Allowed",
          location: url,
          detail:   `Access-Control-Allow-Origin: ${origin}\nAccess-Control-Allow-Credentials: true`,
          severity: "CRITICAL",
          risk:     "Any website can make authenticated requests to your API — reading private data or taking over accounts.",
          fixes: [
            { label: "Whitelist specific origins",
              code:  "// Never combine wildcard with credentials\n// ❌ DANGEROUS\napp.use(cors({ origin: '*', credentials: true }));\n\n// ✅ SAFE\napp.use(cors({ origin: ['https://yoursite.com'], credentials: true }));" }
          ]
        });
      } else if (origin === "*") {
        result.issues.push({
          title:    "CORS Allows All Origins (Wildcard)",
          location: url,
          detail:   "Access-Control-Allow-Origin: * — any website can read this API's responses.",
          severity: "MEDIUM",
          risk:     "Fine for public APIs. Dangerous if any sensitive data is returned.",
          fixes: [
            { label: "Restrict to known origins if response has sensitive data",
              code:  "app.use(cors({ origin: ['https://yoursite.com'] }));" }
          ]
        });
      } else if (origin === EVIL_ORIGIN) {
        result.passed = false;
        result.issues.push({
          title:    "CORS Reflects Arbitrary Origin",
          location: url,
          detail:   `Server reflected evil origin: Access-Control-Allow-Origin: ${origin}`,
          severity: "HIGH",
          risk:     "Server blindly allows any origin — effectively wildcard but works with credentials too.",
          fixes: [
            { label: "Validate origin strictly before reflecting",
              code:  "const allowed = new Set(['https://yoursite.com']);\nif (allowed.has(req.headers.origin)) {\n  res.setHeader('Access-Control-Allow-Origin', req.headers.origin);\n}" }
          ]
        });
      }
    } catch { }
  }

  return result;
}

module.exports = testCORS;
